Support custom tiles source

HarelM opened the pull request:
This PR should facilitate for feature request describe here: Support custom source #29

I've added documentation above the function it self, I don't know how to convert this to proper documentation.
Any help would be appreciated.

As a side note, since I'm using a patch version of mapbox-gl-js I will only be able to help testing (for example current rc) only after this is merged... Sorry...
I would prefer to have all the breaking changes (mainly name changes etc) as soon as possible since I prefer to work in order to migrate to this library only after the breaking changes had occurred to make sure that everything is working once and I won't need to do it in two steps (first only to migrate to this library and the second to make sure I fix things that were broken assuming I want to upgrade to a later version, which I probably will).
This is of-course only my point of view and needs...

nyurik commented on src/util/ajax.js:
```diff
@@ -242,6 +242,12 @@ export const makeRequest = function(requestParameters: RequestParameters, callba
     //   some versions (see https://bugs.webkit.org/show_bug.cgi?id=174980#c2)
     // - Requests for resources with the file:// URI scheme don't work with the Fetch API either. In
     //   this case we unconditionally use XHR on the current thread since referrers don't matter.
+    if(/^custom:/.test(requestParameters.url) && isWorker() && self.worker && self.worker.actor) {
+        return self.worker.actor.send('getResource', requestParameters, callback);
+    }
+    if (/^custom:/.test(requestParameters.url) && !isWorker() && config.LOAD_TILES_FUNCTION != null) {
+        return config.LOAD_TILES_FUNCTION(requestParameters, callback);
+    }
```
i think we wouldn't want to do the same regex test twice, e.g. something like this?  This way no function is ever called multiple times (even if `.isWorker()` and `.test()` are optimized internally)

```suggestion
    if (/^custom:/.test(requestParameters.url)) {
        if (isWorker()) {
            if (self.worker && self.worker.actor) {
                return self.worker.actor.send('getResource', requestParameters, callback);
            }
        } else {
            if (config.LOAD_TILES_FUNCTION != null) {
                return config.LOAD_TILES_FUNCTION(requestParameters, callback);
            }
        }
    }
```

nyurik reviewed (commented):
I am not too sure either way if this is a good approach or not, but just a minor code cleanup suggestion.

snickell commented:
I'd like some time to look over this before we consider it, esp since @HarelM can't test directly (thanks for contributing this @HarelM btw!!!), at the very least, this is probably something we should save for "v1.15".

v1.13: full compat
v1.14: shims and renamed internally
v1.15/v2: consider new features

HarelM commented:
I'll do the requested changes later on, haven't got to it yet.
I guess a possible solution to allow testing this is to push this into a branch and allow branches to be uploaded to npm as a pre-release, or if you have an alternative that would be great too.
My current solution for allowing this in my code is to use the generated js files, place them in the mapbox-gl folder and run `patch-package`.
I can obviously do the same here, but I can't qualify this as proper testing... :-)

nyurik commented:
I have not dug too deep, but it seems a bit weird to introduce a magical `custom` source when what we really need is an easy way to register arbitrary protocols -- it should be trivial for the users of the lib to just call `mapLibreInstance.registerMyProtocol( 'custom', myCallback )`  that would do exactly the same without introducing any magical handling of the custom protocol, plus allow multiple such protocols, plus allow an easy way to override the ones in the system.  I might be missing something though.

HarelM commented:
Good point. I'll see what I can do about it.

HarelM commented:
Thanks for the review! I've pushed changes according it.
The code around this is fairly simple, there's not much to it beside the changes in the `ajax,js` file and setting everything up in `index.js` file, this is what makes it an elegant solution from my point of view.
Let me know if there are other changes required, and let me know when there's a RC with it so I can test it.

HarelM commented:
Recent tests shows that the above code is not working as expected since the `config` object is not defined in the worker context.
The original solution worked since it does not need the config object in the worker context (which is why I think I chose it eventually).
I'll see if I can come up with another solution to this, but I'm not sure I will. If anyone has a thought or idea how to solve this, please let me know...

AbelVM commented:
[SharedArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer) might be the answer, but... [Spectre / Meltdown](https://www.chromium.org/Home/chromium-security/ssca) :disappointed:

tuukka commented:
> If anyone has a thought or idea how to solve this, please let me know...

If we can't access the list of registered custom schemes here, could we instead check that it's a scheme not [handled by `fetch`](https://fetch.spec.whatwg.org/#scheme-fetch)? It would just postpone the error message in the rare case of an unhandled scheme a bit. For example:

```javascript
    if (requestParameters.url.match(/^(\w+):/) &&
        !requestParameters.url.match(/^(about|blob|data|file|ftp|https?):/)
    ) {
        // Use custom handler
    } else {
        // Current logic to use fetch or XMLHttpRequest
    }
```

AbelVM commented:
@HarelM , it looks like the config is accesible from the worker, but the property `LOAD_TILES_FUNCTION_MAP` has the default value. 

![image](https://user-images.githubusercontent.com/9017165/102495477-6cc37700-4076-11eb-84ab-7fe7de313459.png)

Possible reasons I can think of:
1. The  [serializer](https://github.com/IsraelHikingMap/maplibre-gl-js/blob/main/src/util/web_worker_transfer.js) is not working as expected  (nope)
2. The workers are instantiated as soon as the `mapboxgl` module is imported, and therefore, the later config changes are ignored. (nope)
3. The workers import the `config` module, so they will take the default values of the object as declared in the JS file
 
So lucky number 3. [Broadcast Channel API](https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API) would be great to solve this, but it's not supported by Safari. Workers/Actors code region is a bit obscure to me in order to inject some messaging logic to pass the variables :disappointed:

HarelM commented:
I'm guessing sending a function is probably not a good idea, however sending only the keys to allow knowing there's a function in the main thread context is the way to go, IMO.
Having said that, I have no clue how to achieve it yet...
Since I don't expect this to be merged soon I have some time to find out :-)

AbelVM commented:
I've been testing that very approach and no change to config, even adding a plain string, is propagated to the workers. They only see the default values. Even using the original getters and setters to set up API_URL or the token... is not propagated to the workers. I have no idea how on earth they consume those params. 

Maybe [this PR](https://github.com/mapbox/mapbox-gl-js/pull/7594) can shed some light on how can we deal with this issue, and, double combo, get rid of custom Mapbox code

HarelM commented:
I don't think the config is the right approach for the reasons you mentioned in your comment.
There's a need to call dispatch somewhere to pass data to the workers I believe...
I still need to figure out how...

hyperknot commented:
Have you checked out `transformRequest`? It takes a function which transforms all url's before a fetch by mapbox-gl-js.

hyperknot commented:
I'm also using a custom source `maphub:` which I load using `transformRequest` + a regex. It took about 5 lines to implement it using `transformRequest`.

HarelM commented:
@hyperknot I haven't. If you can share a working example, or something that almost works just to get the idea it would be great. If this is already implemented and there's no need to do anything it would be great :-)

HarelM commented:
@hyperknot - this doesn't seem to be a valid solution for reading a tile that is locally stored in a sqlite or indexdb database.
I might be missing out, but I don't see how this solves the case where you don't have network connectivity and would still want to fetch tiles. Please correct me if I'm missing out something obvious...

hyperknot commented:
I don't know the internal details of Cordova, etc. but are you saying that a fetch wouldn't work from an url like object (like file:// or whatever)? If so then indeed transformRequest wouldn't work.

HarelM commented:
I see. So I guess `transformRequest won't work for my needs. You can think of the following scenario which is not related to Cordova. A user clicks on "Make map available offline" button. In that moment downloading of tiles begin and are stored in indexdb (or wherever). When this operation is finished the user should be able to see the map in the browser without network connectivity. Since running fetch for any url won't help there's a need to be able to inject mapbox-gl with the actual data, stored in IndexDb in this case, so a function to support this is needed.

hyperknot commented on src/index.js:
```diff
@@ -170,7 +172,32 @@ const exported = {
+    /**
+     * Sets a custom load tile function that will be called when the using a source that starts with a custom url.
+     * The example below will be triggered for custom:// urls.
+     * The function to be used will recieve the request parameters and should call the callback with the resulting request,
```
recieve -> receive

hyperknot commented on src/util/ajax.js:
```diff
@@ -242,6 +242,15 @@ export const makeRequest = function(requestParameters: RequestParameters, callba
     //   some versions (see https://bugs.webkit.org/show_bug.cgi?id=174980#c2)
     // - Requests for resources with the file:// URI scheme don't work with the Fetch API either. In
     //   this case we unconditionally use XHR on the current thread since referrers don't matter.
+    let key = Object.keys(config.LOAD_TILES_FUNCTION_MAP || {}).find(k => requestParameters.url.startsWith(k));
```
I'm not sure this is a good idea here for each makeRequest. It's a very high performance part of the library and we shouldn't make for loops / find etc. here.

hyperknot replied:
In this regards the hardcoded `custom://` was a better idea, since doing a regex for a simple string is much better then doing this `Object.keys.find(k => startsWith(k))`

nyurik replied:
> In this regards the hardcoded `custom://` was a better idea, since doing a regex for a simple string is much better then doing this `Object.keys.find(k => startsWith(k))`

I agree with you here in the sense that doing a `Object.keys.find()` seems excessive, but it should be fairly trivial to set up a map/object that map protocols->callback, and use something like `registeredProtocols[url.substring(0, url.indexOf('://'))]` -- much more straightforward IMO.

AbelVM commented:
I have a fully working version, should I open a new PR?

HarelM commented:
I looked at the code and I think you can send a PR to my repository and I can merge it, but I think more work is needed here to remove the "file or http" assumption...

AbelVM commented:
The idea is that if not `http(s)://` or `file://`, then it is  a `custom` protocol, so let's manage it. The library itself won't allow you to use any other protocol (but `mapbox://`, obviously).

Looks like the config values are sent to the workers in a "first pass" where the requests are made in the main thread, so the main change is in `ajax.js` file (all other changes are aesthetic ones):

```javascript
if (!(/^https?:|^file:/.test(requestParameters.url))){
        if (isWorker() && self.worker && self.worker.actor) {
            return self.worker.actor.send('getResource', requestParameters, callback);
        }
        if (!isWorker()) {
            const p = requestParameters.url.substring(0, requestParameters.url.indexOf('://'));
            let f = (config.REGISTERED_PROTOCOLS[p] === void 0) ? makeFetchRequest : config.REGISTERED_PROTOCOLS[p];
            return f(requestParameters, callback);
        }
    }
```

If there's not custom function defined for the custom  protocol, it falls back to the default `makeFetchRequest` function. 

Working example: https://stackblitz.com/edit/customsources-v2

HarelM commented:
Yup, I understand what you wrote but I think there might be a way to improve it - i.e. send the custom protocol to the workers and have them update the config they have with the protocol keys (only) and not the function itself.
But it also might be an overkill.
In any case, I think the changes you made are correct and should be in either my PR or a separate one, I honestly don't mind :-)
Further improvement can be done later - this PR should at least work (which is not the case right now...)

AbelVM commented:
Same here, I don't care wich PR is merged as far as it's merged :smile:  

Which option is faster?

![img](https://media.giphy.com/media/U6kI1fjYaaWtoUuiJG/giphy.gif)

HarelM commented:
Both are not going to happen soon enough I guess... :-) I have not maintenance permission here :-)
In order not to loose this thread an whatever, I suggest that you submit the PR to my repo and I'll merge it there so it will appear as part of this PR and not open another one.
My 2 cents, anyway...

AbelVM commented:
Actually, you already have an 1h old PR there :stuck_out_tongue_closed_eyes:  

https://github.com/IsraelHikingMap/maplibre-gl-js/pull/1

hyperknot commented:
@AbelVM you got in some unrelated changes in the PR.

AbelVM commented:
Yes, I just realized, and it's weird. It's a simple fork of this very repo and branch, and I just modified 3 files.

All the `.github/*.md` stuff is not mine, it came with the fork, apparently :thinking:

AbelVM commented:
@HarelM Could you please fix the `extra files` issue in your fork without a PR there to fix the PR here and so on?

HarelM commented:
I'm not sure how to do it... Should I merge from here to my fork in order to solve this?

HarelM commented:
I've tried to manually copy the files from the upstream repository and commit them but when staging them it doesn't seem to show any thing to commit... :-(
In any case looking at the file changes in this PR using the following address removes white space changes:
https://github.com/maplibre/maplibre-gl-js/pull/30/files?w=1
If this is so important I can create a new PR, but honestly, merging this would be simpler and faster...

AbelVM commented:
Sorry for my late reply. I'm having the same issue with those files in a fresh test fork, without even open those files.

Any helping hand?

HarelM commented:
Probably related to check-out windows style and the different line ending. But I don't know how to easily solve this...

AbelVM commented:
Hey @snickell @nyurik this PR is stalled because of some weird git behavior at checking out, could you help us cleaning the offending files and clean the path to greenlight this PR?

Every time I try to fork&pr, I get the same problem with those `.github/*/*.md` files without even opening or committing them

nyurik commented:
I tried to merge current main branch into this PR, but I have no permissions.  Usually it is better to create a branch and submit a PR from it than to submit it from the main. Also, when creating a PR, please click the checkbox "allow maintainers to push changes" - this would allow proj maintainers to push changes directly into your branch.  This is for the future PRs though :)

For now, I pushed merged changes into https://github.com/nyurik/maplibre-gl-js/tree/IsraelHikingMap-main

Please do these commands from your main (make sure you don't have any local changes that are different from what you have submitted here)

```bash
git remote add nyurik https://github.com/nyurik/maplibre-gl-js.git
git fetch nyurik
git reset --hard nyurik/IsraelHikingMap-main
git push
```

this should update this PR.

HarelM commented:
I just did that, looks worse now... :-(

HarelM commented:
@nyurik I added you as maintainer to the IsraelHikingMap repo, feel free to fix anything you see fit.
The comparison between the repositories gives only 5 files but this PR show 94 files :-( and it should be only about three files (readme files should not be a part of this PR)...
Please advise what should I do next, If you want to merge it from your repo I'm fine with it.
If you want me to close this PR and submit a new one that's fine too.
Just let me know what should be the next step, I don't want to create a burden, just want to push this forward...
Thanks for all the hard work!

nyurik commented:
@HarelM git is weird at times. I fixed and force-pushed it, plus added a few minor styling fixes. Note that CI is currently failing your PR due to a few minor errors. Also, note that I have not yet reviewed the actual code, just did a few styling things. I am still a bit worried about the potential for a performance impact, and hope we can make it without any regex text searches (or at least make sure it only runs once, not once per tile).  Regardless, we won't merge this at least until after the initial code "Mapbox compatible" version ships :)

HarelM commented:
Thanks a lot for all your help!! If you have a timeline or rough estimation related to when this would be merged it world be great!
Since this code runs in a worker just before a http request which is a lot more time consuming than a regexp I don't see a real performance impact.
In any case please let me know what exactly you want me to change and I'll fix it.

rbrundritt commented:
I didn't see it in the code, maybe I missed it, but would it make sense to have the request params go through the maps transformRequest function first just in case?

HarelM commented:
Amy updates regarding a timeline for this to be merged? It's sitting here for a long time now...
I'm really waiting for this in order to migrate to maplibre and progress my ability to test this library every time it comes out.
This is really a blocker for me...
The changes are small, it shouldn't break anything, please consider this for the next release...

wipfli commented:
OK let's start by rerunning CI.

AbelVM commented:
I'm impacted by this PR too. 

We are implementing an experimental format (we will share it openly as soon as we release v. 0.1!!), and this PR is the leverage to make it work client-side. In the meantime, we're working with a v1.13 fork just to add this functionality and test our pipeline, but this fork of mine is not prod-ready :disappointed:

wipfli commented:
@AbelVM great, thanks. And you are happy with this code here?

wipfli commented:
@HarelM to rerun CI please push an empty commit which does not really change the code. I cannot rerun it myself because the last commit is older than one month.

AbelVM commented:
> @AbelVM great, thanks. And you are happy with this code here?

It's working :shrug: And the performance is great, the example in https://github.com/maplibre/maplibre-gl-js/pull/30#issuecomment-748611385 performs no transformation at all on the input, but manages any custom format format and just logs the retrieved tile' URL

We're using and old version of this fork, and everything works as expected, or even better. Our internal custom protocol even makes geometry transformations on the fly (and some other dark magic) and it's smooth as silk, as the regex is only evaluated once and everything else runs in workers.

**BUT** There might be some code styling issues, (like `void 0` use, my fault :sweat_smile: , old dog here) or maybe it could be more streamlined with other code pieces in other code regions that I'm not aware of. I just worked in this feature based on @HarelM work trying to solve the multiple protocols support and encapsulating the new feature to limit eventually undesired effects. Feel free to improve it, fix it...!!

wipfli commented:
@AbelVM @HarelM I am happy to hack around to get the CI working. Can you give me write permissions on https://github.com/IsraelHikingMap/maplibre-gl-js/tree/main?

HarelM commented:
@wipfli I think I just did, give or a try and let me know.

wipfli commented:
Linter does not seem happy https://github.com/maplibre/maplibre-gl-js/runs/2749427393?check_suite_focus=true

wipfli commented:
```
/home/runner/work/maplibre-gl-js/maplibre-gl-js/src/util/ajax.js
Error:   245:58  error  Missing space before opening brace           space-before-blocks
Error:   252:28  error  Expected 'undefined' and instead saw 'void'  no-void
```

HarelM commented:
@wipfli you want me to take care of this or would you?

wipfli commented:
I think it is best if you clean up the parts that you propose with this pull request. I am happy to help if the CI produces unrelated problems.

HarelM commented:
OK, sure, no problem, I'm on it. I'll change the typescript definition files while I'm on it...

wipfli commented:
Uh all test have passed. Congrats :)

HarelM commented:
Yay! I made the CI pass. Has to change a bit the regexp to include only urls with `://`. @AbelVM please review my changes.
@wipfli feel free to review and merge.
I also changed the typescript definition file to have this methods and changed it to use maplibregl instead of mapbpxgl.

wipfli commented:
I don't feel comfortable reviewing the actual code you propose here because I don't have the technical knowledge to judge it. But maybe @lseelenbinder, @nyurik, or @JannikGM could share their thoughts?

HarelM commented:
There're not much in this PR honestly, a small change to the ajax.js handling for custom protocols.
In any case, I'd appreciate it very much if someone could review and merge this. I've been waiting patiently for a long time... 
I saw that @thaddmt just pushed a commit to typescript definition file, so it would be nice if he can review the changes I made to the definition file.

bdon commented:
My 2 cents:

I maintain [a fork](https://github.com/protomaps/maplibre-gl-js) to implement a very similar feature, but using an [async transformRequest](https://github.com/protomaps/maplibre-gl-js/commit/916e55529f7a27fb8b36e9a43f8d5ac0c1e78020) as mentioned in this thread above. The downside of the that approach is it's not specific; transformRequest will be called for all HTTP requests, including not only tiles but TileJSON, sprites, and glyph ranges. Also, not all the contexts in which transformRequest is used are asynchronous: [example in the code](https://github.com/maplibre/maplibre-gl-js/blob/main/src/source/load_tilejson.js#L35), so implementing it that way would require refactoring a lot more internals.

I tested out this PR by implementing an adapter for [PMTiles archives](https://github.com/protomaps/PMTiles), a [COG](http://cogeo.org)-like format which fetches raster or vector tiles from a single file via HTTP Range requests. 

[Here is the code of my implementation](https://gist.github.com/bdon/594f05e6137c12b12e76a0aac3d6be8e). You can run this by putting the file in the debug/ directory of your branch checkout. Demo of how it works:

https://user-images.githubusercontent.com/77501/120914601-c6296c00-c6d1-11eb-8a3e-7df26a5c3cef.mp4

## Comments

* Should protocols be specific to tiles and z/x/y or generic to all kinds of resources, including sprites and glyph ranges? Because the current implementation passes an opaque URL string, so you must artificially populate your protocol URL with  `{z} {x} {y}` and then parse it back out with a regex inside your function. [Example on line 26.](https://gist.github.com/bdon/594f05e6137c12b12e76a0aac3d6be8e#file-index-html-L26) If there was a way to pass Z/X/Y integers in the `params`, that would be preferred.

* In my example, I prefixed my resource URL with my protocol name like this: `pmtiles://https://` - I think the original context was to have a `mapbox://` protocol that would assume an HTTPS endpoint, but because this code can't assume HTTP or HTTPS it needs to be awkwardly doubled like this. 

* The protocol function as is can return an error state via the callback as described in [this docstring](https://github.com/maplibre/maplibre-gl-js/pull/30/files#diff-619cc2239a7eb4ef05041767e6ab0e4fb069e98ee9d2c83c2d8e1865c475e141R62): 

```js
} else {                       
   callback(new Error(`Tile fetch error: ${t.statusText}`));                   
}
```
One new condition introduced by protocols is the possibility of missing tiles. For example, if you are fetching tiles from LocalStorage, there might not exist any Tile for a given Z/X/Y, but this is not an error state that should be propagated to the console or other parts of the code. In my implementation [I pass an empty array on line 83](https://gist.github.com/bdon/594f05e6137c12b12e76a0aac3d6be8e#file-index-html-L83) for this case, but I'm not sure if this will behave correctly in all cases. 

## Conclusion

Overall I think this PR is great and minimal, and in my case would discourage me from maintaining a separate fork for the functionality I need, even though the developer ergonomics aren't perfect. It is difficult to predict what protocols others are going to implement, and a more full-fledged API could be evolved from the common pain points shared among Protocols. I think there's some aversion to merging a new feature like this because of the implications of backward compatibility later - maybe it would a good precedent to mark `addProtocol` as an "EXPERIMENTAL" api that developers should be careful about when upgrading versions?

👍

AbelVM commented:
> Yay! I made the CI pass. I had to change a bit the regexp to include only urls with `://`. @AbelVM please review my changes.
> @wipfli feel free to review and merge.
> I also changed the typescript definition file to have this methods and changed it to use `maplibregl` instead of `mapbpxgl`.

It feels right to me :slightly_smiling_face: We are getting closer to get this functionality onboard! :tada: 

Regarding @bdon comments, I think the (minimal) working example I posted at https://github.com/maplibre/maplibre-gl-js/pull/30#issuecomment-748611385 manages all the bullet points you mentioned :nerd_face: . Regarding the need of `artificially populate your protocol URL with {z} {x} {y}...`, just check this PR title: its only target is to provide custom **tiled** sources, so the source' tiles URL must be as expected by the standard, and then you manage those values within the protocol definition.

Joxit reviewed (approved):
The renaming (mapboxgl to maplibregl) is a bit noisy in this PR, maybe this should be done in a separate PR?

Otherwise LGTM

HarelM commented:
If the typescript definition is updated in main I can merge and resolve the conflicts to allow the PR to be less noisy.
Opening another PR, waiting for it to be merged and only then merge this one seems a bit too much (for me...) just for the typings (which were incorrect and I needed to update them anyway).
Regarding experimental suggestion: we can leave it out of the current documentation and see how it goes, if we see that it's good enough as is it can enter or be refactored.
I OK with either approach, I mainly want this in so I can migrate my project.
Once merged, if this can be published as rc I can start progressing my code and project, and probably also ngx-mapbox (which I use) to this org so anyone working with angular (and maybe ionic) can benefit from this as well.
Ping me.
@Joxit Thanks for the review! I appreciate it.

lseelenbinder commented:
I like this implementation, but I agree with @bdon that it should be noted that it's experimental and subject to change for a few iterations.

@bdon We could suggest replacing something like `pmtiles://https://` with `pmtiles+https://` similar to how git uses various transports? Obviously parsing would need to be handled in the function, so it's not ideal. We could also adjust this PR to support `<protocol>+<transport>://` and parse out those two details for the callback.

HarelM commented:
I'm good with everything you decide, just let me know...
There's also an option to merge this and later on improve this as was suggested here according to developer needs/requirements...
I don't know how to mark is as experimental, so please advise on the proper way (although I think that not documenting it would be the definition of experimental, IMHO).
Seems that everything on the checklist is green, please let me know if and when this is planned to be merged.
Thanks guys for the awesome work!

wipfli commented:
Since `addProtocol` will be part of the public API I would prefer to have documentation for what it does.

HarelM commented:
@Joxit I've created the following PR to address the "noise": #172.
Please review it and merge it here so I can carry it downsteam to my repo and reduce the code related to this PR, or simply merge this PR and I'll close the other one. Your call...

AbelVM commented:
Looks like Looks like merging https://github.com/maplibre/maplibre-gl-js/pull/172 has tagged this PR as `dirty` :disappointed: 

Once this is fixed, is there any way to help it move forward and be approved?

HarelM commented:
I'm planning to fix this tonight unless you @AbelVM want to do it first (I.e. Merge main to branch and resolve conflicts).
I'm not sure what's the latest decision regarding documentation, experimental etc.
I favor experimental, and without documentation at this point.
Once we get some feedback we can decide on our the next steps.
So to summarize, I'm in favor of merging as is and releasing a rc version.
Any objections?

HarelM commented:
I've reduced the "noise" last night as requested by @Joxit.
@lseelenbinder I'm not sure how to mark something as experimental, please advise.
Also, please see my previous comment.
I've also sent a message in Slack, I'm planning on merging this tonight (in about 10 hours) if there are not objections...
Worst case this can be reverted as there's not much code involved in this feature...

Joxit commented:
Thank you for your changes! 😁
LGTM

AbelVM commented:
Hi, @snickell , any chance to green-light this PR soon? :sweat_smile:

lseelenbinder commented:
Thanks @HarelM! I think we can just make a note in the documentation that it's subject to change, but it's also so simple I think it may not really matter.
